pkg/ha/state/lease.go

```diff
@@ -69,44 +75,117 @@ func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT,
+	}
+
+	if !maxT.After(currentDeadline) {
+		return false, time.Time{}, nil
+	}
+
+	if err := l.TryChangeLeader(currT); err != nil {
+		return false, time.Time{}, err
+	}
+	return l.confirmLeaseValidation(replica, minT)
+}
+
+func (l *Lease) confirmLeaseValidation(replica string, minT time.Time) (bool, time.Time, error) {
+	l._mu.RLock()
+	defer l._mu.RUnlock()
+
+	if l.leader != replica {
+		return false, time.Time{}, nil
+	}
+
+	if minT.Before(l.leaseStart) {
+		minT = l.leaseStart
+	}
+	return true, minT, nil
+}
+
+func (l *Lease) RefreshLease() error {
+	l._mu.RLock()
+	leader := l.leader
+	leaseStart := l.leaseStart
+	maxTimeSeenLeader := l.maxTimeSeenLeader
+	l._mu.RUnlock()
+	return l.updateLease(leader, leaseStart, maxTimeSeenLeader)
+}
+
+// updateLease uses the supplied client to attempt to update the lease for the potentialLeader.
 // It either updates the lease, or a new leader with the assigned lease interval is set, as
 // signified by the database as the source of truth.
 // An error is returned if the db can't be reached.
-func (h *Lease) UpdateLease(c client.LeaseClient, potentialLeader string, minT, maxT time.Time) (*LeaseView, error) {
+func (h *Lease) updateLease(potentialLeader string, minT, maxT time.Time) error {
 	h._mu.RLock()
 	cluster := h.cluster
 	if h.leader != potentialLeader {
 		h._mu.RUnlock()
-		return nil, fmt.Errorf("should never be updating the lease for a non-leader")
+		return fmt.Errorf("should never be updating the lease for a non-leader")
 	}
 	h._mu.RUnlock()
-	stateFromDB, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	stateFromDB, err := h.client.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
 	if err != nil {
-		return nil, fmt.Errorf("could not update lease from db: %#v", err)
+		return fmt.Errorf("could not update lease from db: %#v", err)
 	}
-	return h.setUpdateFromDB(stateFromDB), nil
+	h.setUpdateFromDB(stateFromDB)
+	return nil
 }
 
 // TryChangeLeader uses the supplied client to attempt to change the leader
 // of the cluster based on the maximum observed data time and the instance
 // that had it. If updates the lease with the latest state from the database.
 // An error is returned if the db can't be reached.
-func (h *Lease) TryChangeLeader(c client.LeaseClient) (*LeaseView, error) {
-	h._mu.RLock()
-	cluster := h.cluster
-	maxTimeInstance := h.maxTimeInstance
-	maxTimeSeen := h.maxTimeSeen
-	h._mu.RUnlock()
-	leaseState, err := c.TryChangeLeader(
-		context.Background(), cluster, maxTimeInstance, maxTimeSeen,
+func (l *Lease) TryChangeLeader(currT time.Time) error {
+	l._mu.RLock()
+	cluster := l.cluster
+	currentDeadline := l.leaseUntil
+	maxTimeInstance := l.maxTimeInstance
+	maxTimeSeen := l.maxTimeSeen
+	recentLeaderSeen := l.recentLeaderWriteTime
+	l._mu.RUnlock()
+
+	if currentDeadline.After(maxTimeSeen) {
+		return nil
+	}
+
+	if currT.Sub(recentLeaderSeen) <= haLastWriteInterval {
```

Comment:
can we add a test that would catch the bug we had here before?

Reply:
Sure, in a separate commit.

Reply:
@antekresic  can I know more info on the bug we noticed here? I am curious to know how did I not find it in my manual tests...

Reply:
Ahh, just noticed it here https://github.com/timescale/promscale/pull/569